src/hooks/usePrivateChats.ts

```diff
@@ -0,0 +1,184 @@
+import { useCallback, useMemo } from "react";
+
+import { sendPrivateMessage, setChatMessageIsRead } from "api/chat";
+
+import { myPrivateChatsSelector } from "utils/selectors";
+import {
+  chatSort,
+  buildMessage,
+  getPreviewChatMessageToDisplay,
+  getMessageToDisplay,
+} from "utils/chat";
+import { WithId } from "utils/id";
+
+import { PreviewChatMessage, PrivateChatMessage } from "types/chat";
+
+import { isLoaded, useFirestoreConnect } from "./useFirestoreConnect";
+import { useSelector } from "./useSelector";
+import { useUser } from "./useUser";
+import { useWorldUsersById } from "./users";
+
+export type ChatToDisplay = {};
+
+export const useMyPrivateChatsConnect = () => {
+  const { user } = useUser();
+
+  useFirestoreConnect(() => {
+    if (!user?.uid) return [];
+
+    return [
+      {
+        collection: "privatechats",
+        doc: user.uid,
+        subcollections: [{ collection: "chats" }],
+        storeAs: "myPrivateChats",
+      },
+    ];
+  });
+};
+
+export const useMyPrivateMessages = () => {
+  useMyPrivateChatsConnect();
+
+  const myPrivateMessages = useSelector(myPrivateChatsSelector);
+
+  return useMemo(
+    () => ({
+      myPrivateMessages: myPrivateMessages ?? [],
+      isUserPrivateChatsLoaded: isLoaded(myPrivateMessages),
+    }),
+    [myPrivateMessages]
+  );
+};
+
+export const usePrivateChatList = () => {
+  const { user } = useUser();
+  const { worldUsersById } = useWorldUsersById();
+  const { myPrivateMessages } = useMyPrivateMessages();
+
+  const userId = user?.uid;
+
+  const recentPreviewChatMessagesMap = useMemo(
+    () =>
+      myPrivateMessages.reduce<{
+        [key: string]: PreviewChatMessage;
+      }>((acc, message) => {
+        if (!userId) return acc;
+
+        const { from: fromUserId, to: toUserId } = message;
+
+        // NOTE: Either `from` author or `to` author is Me. Filter me out
+        const counterPartyUserId =
+          fromUserId === userId ? toUserId : fromUserId;
+
+        // NOTE: Filter out not existent users
+        if (!worldUsersById[counterPartyUserId]) return acc;
+
+        if (counterPartyUserId in acc) {
+          const previousMessage = acc[counterPartyUserId];
+
+          // NOTE: If the message is older, replace it with the more recent one
+          if (previousMessage.ts_utc > message.ts_utc) return acc;
+
+          return {
+            ...acc,
+            [counterPartyUserId]: { ...message, counterPartyUserId },
+          };
+        }
+
+        return {
+          ...acc,
+          [counterPartyUserId]: { ...message, counterPartyUserId },
+        };
```

Comment:
Could extract the logic of forming this object into a function, but tbh looking at it, I wouldn't say we can benefit much from it. I'm open to suggestions though because low-key I feel the same

Reply:
At the very least it'd probably be worth making a little inline function somewhere within that block.. probably within the `useMemo`, but outside of the `reduce` That'll simplify/clean up the code a bit at the very least.